### PR TITLE
Sync column and table comments

### PR DESCRIPTION
**Pull request by EvilPuppetMaster**
Hi,

Here is an implementation of the feature to use the COMMENTs on tables and fields to populate the metabase descriptions when syncing.
Related issue: https://github.com/metabase/metabase/issues/3089

This is my first time using Clojure so I'm sure I've committed many faux pas. Feedback is welcomed! 

Note I've only got unit tests against H2 and Postgres. Happy to test against other DBs if you can point me to docs about how to do so. I definitely will test against Presto and SQL Server, which are the DBs I need! But I also didn't want to do too much of that without first getting a code review and a general thumbs up for the approach.

Cheers
Tom

###### TODO 
-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

### CODE REVIEW

File: test/metabase/test/data/h2.clj
Comment (camsaul) [inline test/metabase/test/data/h2.clj]: ⁉️

File: test/metabase/test/data/generic_sql.clj
Comment (camsaul) [inline test/metabase/test/data/generic_sql.clj]: normally in Clojure a boolean parameter to a function gets a `?` at the end to mark it as such. `skip-drop-db?` would make more sense here

File: test/metabase/test/data/generic_sql.clj
Comment (camsaul) [inline test/metabase/test/data/generic_sql.clj]: For conditionals that don't have an `else` clause you should just use `when`/`when-not` instead of `if`/`if-not`. https://github.com/bbatsov/clojure-style-guide#when-not-instead-of-single-branch-if-not

File: test/metabase/test/data/generic_sql.clj
Comment (camsaul) [inline test/metabase/test/data/generic_sql.clj]: `(str "")` is redundant. You can just do `""`. It would probably make more sense instead to just have this function return `nil` instead of a blank string anyway

File: test/metabase/test/data/generic_sql.clj
Comment (camsaul) [inline test/metabase/test/data/generic_sql.clj]: Same thing applies here -- `(str "")` is `""` -- but just returning nothing instead probably makes more sense

File: test/metabase/test/data/generic_sql.clj
Comment (camsaul) [inline test/metabase/test/data/generic_sql.clj]: same thing here (redundant call to `str`)

File: test/metabase/test/data/generic_sql.clj
Comment (camsaul) [inline test/metabase/test/data/generic_sql.clj]: same thing here (redundant call to `str`)

File: test/metabase/test/data/generic_sql.clj
Comment (camsaul) [inline test/metabase/test/data/generic_sql.clj]: probably better just to make these all be `(constantly nil)` rather than blank strings and then have the function that actually uses their values check to see if it gets something back before using the value

Comment (camsaul) [reply test/metabase/test/data/generic_sql.clj]: otherwise someone could implement this function for a database, and if it were to return `nil` it would cause a `NullPointerException` in the call to `format` where you actually use it

File: test/metabase/test/data/generic_sql.clj
Comment (camsaul) [reply test/metabase/test/data/generic_sql.clj]: e.g.

```clojure
(when (seq field-comment)
   (format "COMMENT '%s'" field-comment))
```

Note the use of `seq` above, which is called nil-punning, to make sure `field-comment` is non-blank (`(seq "")` is `nil`).

File: src/metabase/sync/sync_metadata/tables.clj
Comment (camsaul) [inline src/metabase/sync/sync_metadata/tables.clj]: This looks like it will do an `UPDATE` for every single "changed" table regardless of whether it has a description or not.

On a first sync of a DB with 100,000 tables and 0 table comments that would result in an extra 100,000 DB calls like `UPDATE table SET description = NULL WHERE ...`. We *definitely* need to make sure we're not doing DB calls here to set a column that's already `NULL` to `NULL`.

File: src/metabase/driver/generic_sql.clj
Comment (camsaul) [inline src/metabase/driver/generic_sql.clj]: https://github.com/bbatsov/clojure-style-guide#when-not

File: src/metabase/driver/generic_sql.clj
Comment (camsaul) [inline src/metabase/driver/generic_sql.clj]: https://github.com/bbatsov/clojure-style-guide#when-not

Comment (camsaul) [reply src/metabase/driver/generic_sql.clj]: It should be fine to do 

```clojure
{:name        (:table_name table)
 :schema      schema
 :description (when-not (str/blank? remarks)
                remarks)}
```

since we should be checking whether `description` is `nil` anyway before saving it

File: src/metabase/sync/sync_metadata/fields.clj
Comment (camsaul) [inline src/metabase/sync/sync_metadata/fields.clj]: this logic (checking to make sure `new-field-comment` isn't blank) is the sort of thing we need to have for Table as well

File: src/metabase/sync/sync_metadata/fields.clj
Comment (camsaul) [inline src/metabase/sync/sync_metadata/fields.clj]: What's the reasoning behind this change?

File: test/metabase/test/data/h2.clj
Comment (camsaul) [inline test/metabase/test/data/h2.clj]: it would be nice to keep everything indented nicely

**Review by camsaul — CHANGES_REQUESTED** [review]
Overall this looks really good. Thanks for the contribution.

There are a couple changes we need to make before merging. The big one is making sure we're not doing an extra `UPDATE` call that doesn't do anything for every table. Besides that there are some style nitpicks and other small suggestions. Once those are cleaned up I think we can merge this

**Comment by EvilPuppetMaster** [comment]
Thanks @camsaul for the feedback and the style tips. Your suggested changes are made.

I haven't yet managed to test on DBs other than h2 / postgres, but I can't see any detrimental effects of merging it despite that. It's relying on a standard feature of the JDBC API. I think the worst that could happen is that some DBs don't support that, eg Oracle as you pointed out in the related issue https://github.com/metabase/metabase/issues/3089, but I believe all that will happen in such cases is the remarks will simply be null which is a handled case anyway, and comments won't be supported for those DBs until someone (perhaps me in a future PR) adds tests for those DBs and deals with any DB specific quirks.

**Comment by EvilPuppetMaster** [comment]
Hi @camsaul,

I've just checked on this for the first time in a while and noticed there are now conflicts. I can fix them up but it would be good to know that it can be merged soon if I do? 

Cheers
Tom

**Comment by sabino** [comment]
@salsakran @camsaul do you think it is possible to have this merged anytime soon?

In a near future maybe, could be nice to add the possibility to make global search consider the schemas, tables, columns and comments as well, but thats another idea.

**Comment by senior** [comment]
@EvilPuppetMaster Thank you for your work on this PR! Currently only branches on the Metabase central repository (not PRs) trigger CI for our supported backends. I've taken your commits, merged in the latest code from master and submitted it as https://github.com/metabase/metabase/pull/6974. I will work on getting this merged and included in `0.31.0` which should have a release candidate build out soon.

**Comment by EvilPuppetMaster** [comment]
@senior That's excellent news thank you? I take it you resolved the merge conflicts that existed?

**Pull request merged**
